src/core/job.c

```diff
@@ -949,22 +950,43 @@ static int job_dispatch_timer(sd_event_source *s, uint64_t monotonic, void *user
         return 0;
 }
 
-int job_start_timer(Job *j) {
+
+int job_start_timer(Job *j, bool job_running) {
         int r;
+        usec_t run_begin, timeout_time;
 
-        if (j->timer_event_source)
-                return 0;
+        if (job_running) {
+                if (j->unit->job_running_timeout == USEC_INFINITY)
+                        return 0;
 
-        j->begin_usec = now(CLOCK_MONOTONIC);
+                run_begin = now(CLOCK_MONOTONIC);
 
-        if (j->unit->job_timeout == USEC_INFINITY)
-                return 0;
+                if (j->timer_event_source) {
+                        /* Update only if JobRunningTimeout results in earlier timeout */
+                        (void) sd_event_source_get_time(j->timer_event_source, &timeout_time);
+                        if (timeout_time <= run_begin + j->unit->job_running_timeout)
+                                return 0;
+                        r = sd_event_source_set_time(j->timer_event_source, run_begin + j->unit->job_running_timeout);
```

Comment:
please use usec_add() for adding timeouts, it deals somewhat sanely with overflows